### DISCUSSION TITLE
How to listen to the parsed result immediately?

### DISCUSSION THREAD

**Discussion by dhaval15** [petitparser/dart-petitparser#110] [Q&A] [answered]
I want to parse file that has metadata on top, and I want to parse that metadata and populate that data to parse the rest of file ? How would I do that. As far as I understand the map function is called on second pass after successful parsing. So with MapParser I couldn't achieve that or am I missing something ?

**Comment by renggli** [comment, marked as answer]
Probably the easiest is if you split this into two separate parsing steps: You can use the `Result` of the meta-data parser to build the data parser, as well as to resume the parse where the meta-data parser completed. Something along these lines:

```dart
final metadataResult = metadataParser.parse(input);
final dataParser = buildDataParser(metadataResult.value);
final dataResult = dataParser.parseOn(metadataResult);
```

Alternatively you can use the [ContinuationParser](https://github.com/petitparser/dart-petitparser/blob/main/petitparser/lib/src/parser/action/continuation.dart) to merge this into a single parser that returns the result of the data parser:

```dart
final parser = metadataParser.callCC((continuation, context) {
  final metadataResult = continuation(context);
  final dataParser = buildDataParser(metadataResult.value);
  return dataParser.parseOn(metadataResult);
});
```

**Comment by renggli** [comment]
To also answer your question about the map function: It is immediately called if the inner parser succeeds. There is no second pass, everything is done in a single pass.

**Comment by dhaval15** [comment]
> Probably the easiest is if you split this into two separate parsing steps: You can use the `Result` of the meta-data parser to build the data parser, as well as to resume the parse where the meta-data parser completed. Something along these lines:
> 
> ```dart
> final metadataResult = metadataParser.parse(input);
> final dataParser = buildDataParser(metadataResult.value);
> final dataResult = dataParser.parseOn(metadataResult);
> ```
> 
> Alternatively you can use the [ContinuationParser](https://github.com/petitparser/dart-petitparser/blob/main/petitparser/lib/src/parser/action/continuation.dart) to merge this into a single parser that returns the result of the data parser:
> 
> ```dart
> final parser = metadataParser.callCC((continuation, context) {
>   final metadataResult = continuation(context);
>   final dataParser = buildDataParser(metadataResult.value);
>   return dataParser.parseOn(metadataResult);
> });
> ```

Thanks, the first approach worked.
